service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java

```diff
@@ -86,6 +86,8 @@
 @Builder(toBuilder = true)
 @AllArgsConstructor
 @HasDocumentsIncludedInSwet(groups = UploadDocumentsGroup.class)
+@TimeRange(message = "Date must be within 8 days of the order date", groups = EPOEndDateGroup.class,
```

Comment:
this is not good idea, placing such annotation at class level does not give a clue which filed(s) it validates

Reply:
i see the problem, maybe annotation should be named EPOTimeRange

Reply:
other option, which involves reflection could be like;

`
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = TimeRangeValidator.class)
public @interface TimeRange {
    String message() default "Enter a valid time";
    Class<?>[] groups() default {};
    Class<? extends Payload>[] payload() default {};

    TimeDifference maxDate() default @TimeDifference(amount = 0, unit = SECONDS);
    
    String dateMethod();
}

and then

`
@Component
public class TimeRangeValidator implements ConstraintValidator<TimeRange, CaseData> {

    private TimeDifference maxDate;
    private String dateMethod;

    @Override
    public void initialize(TimeRange annotation) {
        maxDate = annotation.maxDate();
        dateMethod = annotation.dateMethod();
    }

    @SneakyThrows
    @Override
    public boolean isValid(CaseData caseData, ConstraintValidatorContext context) {
        
        final LocalDateTime startDate;
        
        if(ObjectUtils.isEmpty(dateMethod)){
            startDate = LocalDateTime.now();
        }else {
            startDate = (LocalDateTime) ReflectionUtils.findMethod(caseData.getClass(), dateMethod).invoke(caseData);
        }
        //LocalDateTime startDate = caseData.getDateAndTimeOfIssue();
        LocalDateTime rangeEnd = startDate.plus(maxDate.amount(), maxDate.unit());

        return !caseData.getEpoEndDate().isAfter(rangeEnd);
    }
}
`
`

Reply:
lets not bother with reflection, as you mentioned it can have some problems at compile time